File: sdk/src/test/java/eu/darken/porter/sdk/PorterProtocolWireUserServiceTest.java
package eu.darken.porter.sdk;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;

import android.content.ComponentName;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import eu.darken.porter.server.IPorterServiceConnection;

/** The binder identity a user service binding presents to the server. */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class PorterProtocolWireUserServiceTest {

    private static final String PACKAGE = "eu.darken.porter.probe";
    private static final String CLASS = "ProbeService";
    private static final String PROCESS_SUFFIX = "probe";

    /** Records the connection binder of every add and remove the server was asked for. */
    private static class RecordingService extends FakePorterService {

        final List<IBinder> connections = Collections.synchronizedList(new ArrayList<>());

        @Override
        public int addUserService(IPorterServiceConnection conn, Bundle args) {
            connections.add(conn == null ? null : conn.asBinder());
            return super.addUserService(conn, args);
        }

        @Override
        public int removeUserService(IPorterServiceConnection conn, Bundle args) {
            connections.add(conn == null ? null : conn.asBinder());
            return super.removeUserService(conn, args);
        }
    }

    /**
     * Holds its first lookup until the other thread reaches one too, so a wire that looks up and
     * registers without a lock creates two stubs for the one callback.
     */
    private static final class GatedCallback implements UserServiceCallback {

        private final Map<PorterBackend, IBinder> registered = new ConcurrentHashMap<>();
        private final CountDownLatch lookups = new CountDownLatch(2);

        @Nullable
        @Override
        public IBinder registeredBinder(@NonNull PorterBackend backend) {
            lookups.countDown();
            try {
                lookups.await(200, TimeUnit.MILLISECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            return registered.get(backend);
        }

        @Override
        public void rememberRegisteredBinder(@NonNull PorterBackend backend, @NonNull IBinder binder) {
            registered.put(backend, binder);
        }

        @Override
        public void connected(@NonNull IBinder binder) {
        }

        @Override
        public void died() {
        }
    }

    private static final ServiceConnection NO_OP = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
        }
    };

    @After
    public void teardown() {
        Porter.resetForTest();
    }

    private static RecordingService attached() {
        RecordingService fake = new RecordingService();
        Porter.onBinderReceived(fake, PACKAGE);
        return fake;
    }

    private static Porter.UserServiceArgs args(String tag) {
        return new Porter.UserServiceArgs(new ComponentName(PACKAGE, CLASS))
                .processNameSuffix(PROCESS_SUFFIX)
                .tag(tag);
    }

    @Test
    public void oneBindingPresentsOneBinderAcrossAddAndRemove() {
        RecordingService fake = attached();
        Porter.UserServiceArgs args = args("one-binding");

        Porter.bindUserService(args, NO_OP);
        Porter.unbindUserService(args, NO_OP, false);

        assertEquals(2, fake.connections.size());
        assertNotNull(fake.connections.get(0));
        assertSame(fake.connections.get(0), fake.connections.get(1));
    }

    @Test
    public void twoBindingsPresentTwoBinders() {
        RecordingService fake = attached();

        Porter.bindUserService(args("first-binding"), NO_OP);
        Porter.bindUserService(args("second-binding"), NO_OP);

        assertEquals(2, fake.connections.size());
        assertNotNull(fake.connections.get(0));
        assertNotNull(fake.connections.get(1));
        assertNotSame(fake.connections.get(0), fake.connections.get(1));
    }

    @Test
    public void removingAndKillingNamesNoConnection() {
        RecordingService fake = attached();

        Porter.unbindUserService(args("killed-binding"), NO_OP, true);

        assertEquals(1, fake.connections.size());
        assertNull(fake.connections.get(0));
    }

    @Test
    public void twoThreadsBindingOneServiceRegisterOneBinder() throws Exception {
        RecordingService fake = attached();
        GatedCallback callback = new GatedCallback();
        Bundle add = args("concurrent-binding").forAdd();

        Runnable bind = () -> Porter.requireWire().addUserService(callback, add);
        Thread first = new Thread(bind);
        Thread second = new Thread(bind);
        first.start();
        second.start();
        first.join();
        second.join();

        assertEquals(2, fake.connections.size());
        assertNotNull(fake.connections.get(0));
        assertSame(fake.connections.get(0), fake.connections.get(1));
    }
}

File: sdk/src/test/java/eu/darken/porter/sdk/PorterProtocolWireUserServiceTest.kt
package eu.darken.porter.sdk

import android.os.IBinder
import eu.darken.porter.sdk.UserServiceTestSupport.args
import eu.darken.porter.sdk.UserServiceTestSupport.connection
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.test.runTest
import org.junit.After
import org.junit.Before
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNotSame
import org.junit.Assert.assertNull
import org.junit.Assert.assertSame
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.Dispatchers

/** The binder identity a user service binding presents to the server. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
internal class PorterProtocolWireUserServiceTest {

    /**
     * Holds its first lookup until the other thread reaches one too, so a wire that looks up and
     * registers without a lock creates two stubs for the one callback.
     */
    private class GatedCallback : UserServiceCallback {

        private val registered = ConcurrentHashMap<PorterBackend, IBinder>()
        private val lookups = CountDownLatch(2)

        override fun registeredBinder(backend: PorterBackend): IBinder? {
            lookups.countDown()
            try {
                lookups.await(200, TimeUnit.MILLISECONDS)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }
            return registered[backend]
        }

        override fun rememberRegisteredBinder(backend: PorterBackend, binder: IBinder) {
            registered[backend] = binder
        }

        override fun connected(binder: IBinder) {
        }

        override fun died() {
        }
    }

    @Before
    fun inlineServerCalls() {
        // Server calls run inline, so each step below has happened when the next one asserts.
        Porter.ioDispatcher = Dispatchers.Unconfined
    }

    @After
    fun teardown() {
        Porter.resetForTest()
    }

    private fun attached(): ScriptedPorterService {
        val fake = ScriptedPorterService()
        Porter.onBinderReceived(fake, UserServiceTestSupport.PACKAGE)
        return fake
    }

    @Test
    fun oneBindingPresentsOneBinderAcrossAddAndRemove() = runTest {
        val fake = attached()
        val args = args("one-binding")

        val collector = RecordingCollector(backgroundScope, connection().userService(args))
        collector.job.cancelAndJoin()

        assertEquals(2, fake.connections.size)
        assertNotNull(fake.connections[0])
        assertSame(fake.connections[0], fake.connections[1])
    }

    @Test
    fun twoBindingsPresentTwoBinders() = runTest {
        val fake = attached()

        RecordingCollector(backgroundScope, connection().userService(args("first-binding")))
        RecordingCollector(backgroundScope, connection().userService(args("second-binding")))

        assertEquals(2, fake.connections.size)
        assertNotNull(fake.connections[0])
        assertNotNull(fake.connections[1])
        assertNotSame(fake.connections[0], fake.connections[1])
    }

    @Test
    fun removingAndKillingNamesNoConnection() = runBlocking<Unit> {
        val fake = attached()

        connection().stopUserService(args("killed-binding"))

        assertEquals(1, fake.connections.size)
        assertNull(fake.connections[0])
    }

    @Test
    fun twoThreadsBindingOneServiceRegisterOneBinder() {
        val fake = attached()
        val callback = GatedCallback()
        val add = args("concurrent-binding")

        val wire = connection().wire
        val bind = Runnable { wire.addUserService(callback, add, false) }
        val first = Thread(bind)
        val second = Thread(bind)
        first.start()
        second.start()
        first.join()
        second.join()

        assertEquals(2, fake.connections.size)
        assertNotNull(fake.connections[0])
        assertSame(fake.connections[0], fake.connections[1])
    }
}
